### ISSUE TITLE
Mypy fails to correctly validate Type[T]

### ISSUE THREAD

**Issue by BvB93** [python/mypy#8681]
Dear all,

Consider the example code below:

``` python
from typing import Type, TypeVar, Tuple

T = TypeVar('T')


def func(a: T, a_type: Type[T]) -> Tuple[T, Type[T]]:
    return a, a_type


a1, a1_type = func(1, int)
a2, a2_type = func(1, str)

reveal_type(a1)
reveal_type(a1_type)
reveal_type(a2)
reveal_type(a2_type)
```

The first call to ``func`` shouldn't give rise to any issues, while the second one should complain about an incompatible type in the ``a_type`` argument (as ``type(1) is not str``).

Yet, whenever I run ``mypy`` (Python 3.7; mypy 0.770; MacOS 10.15.1) it fails to produce any errors, even going as far to infer the type of ``a2`` as ``object`` instead of ``int``. 

```  python
>>> mypy test.py
test.py:13: note: Revealed type is 'builtins.int'
test.py:14: note: Revealed type is 'Type[builtins.int]'
test.py:15: note: Revealed type is 'builtins.object'
test.py:16: note: Revealed type is 'Type[builtins.object]'
```

Is anyone familiar with this issue?

**Comment by JelleZijlstra** [comment]
Yes, in cases like this mypy infers the typevar as the common supertype of the two types. Is this causing you problems in concrete code?

A way to get this to work as you expect could be to define the typevar as `TypeVar("T", int, str)`, but of course then it would not accept other types.

**Comment by BvB93** [comment]
The concrete problem I'm facing is in a wrapper function around the likes of around an assertion function, where ``a`` should always be an instance of ``a_type``.

``` python
def assert_isinstance(a: T, a_type: Type[T]) -> None:
    ...
    assert isinstance(a, a_type)
    ...
```

It's always possible to annotate the arguments simply using ``Any`` and ``type``, but this also means that the annotations are quite a bit less descriptive (from a typing standpoint).

Still, I do find it somewhat questionable that mypy defaults to ``T == object`` in such a case, as this can effectively makes a ``TypeVar`` borderline useless in a rather (dangerously silent) manner.
